### DISCUSSION TITLE
"Undefined reference" when trying to reference Rust libraries from C.

### DISCUSSION THREAD

**Discussion by IoIxD** [corrosion-rs/corrosion#561] [Q&A] [answered]
Been scratching my head for hours wondering what I'm doing wrong. I've used this successfully with other projects, but for some reason, I can't use it with this one without getting a bunch of errors:

```
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::DynamicImage(std::vector<char, std::allocator<char> >)':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:4:(.text+0x39): undefined reference to `internal__Load__DynamicImage__FromMemory(unsigned char*, unsigned long)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::GetDimensions()':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:8:(.text+0x5e): undefined reference to `internal__DynamicImage__Dimensions(__DynamicImage__*)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::GetPixel(unsigned int, unsigned int)':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:11:(.text+0x86): undefined reference to `internal__DynamicImage__GetPixel(__DynamicImage__*, unsigned int, unsigned int)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::width()':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:14:(.text+0xa0): undefined reference to `internal__DynamicImage__GetWidth(__DynamicImage__*)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::height()':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:17:(.text+0xba): undefined reference to `internal__DynamicImage__GetHeight(__DynamicImage__*)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::InBounds(unsigned int, unsigned int)':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:20:(.text+0xe2): undefined reference to `internal__DynamicImage__InBounds(__DynamicImage__*, unsigned int, unsigned int)'
/usr/bin/ld: CMakeFiles/ligma-browser.dir/imageload/imageload.cpp.o: in function `DynamicImage::~DynamicImage()':
/home/gavin/Projects/ligma-browser/imageload/imageload.cpp:23:(.text+0xfc): undefined reference to `internal__DynamicImage__Free(__DynamicImage__*)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/ligma-browser.dir/build.make:109: ligma-browser] Error 1
make[1]: *** [CMakeFiles/Makefile2:351: CMakeFiles/ligma-browser.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

The relevant code is:
lib.rs:
```rs lib.rs
#[no_mangle]
pub extern "C" fn internal__Load__DynamicImage__FromMemory(
    data: *mut u8,
    len: usize,
) -> __DynamicImage__ {
    let slice = unsafe { std::slice::from_raw_parts(data, len) };
    let image = load_from_memory(slice).unwrap();

    __DynamicImage__ {
        image: Box::leak(Box::new(image)) as *mut image::DynamicImage as *mut c_void,
    }
}
```
Which is fed into cbindgen, resulting in the following __internal__imageload.hpp:
```c __internal__imageload.hpp
struct __DynamicImage__ internal__Load__DynamicImage__FromMemory(uint8_t *data, uintptr_t len);
```
Referenced by imageload.cpp:
```cpp
DynamicImage::DynamicImage(std::vector<char> data) {
  this->img = internal__Load__DynamicImage__FromMemory((uint8_t*)data.data(),
                                                       data.size());
};
```
The function in question being what actually throws the linker error.

I'm only using the library as I have in the past.

[cmake output](https://www.toptal.com/developers/hastebin)

And if you must compile, the cmake file handles the dependencies automatically, just make sure you have glm, GLEW, and libcurl.

**Comment by jschwe** [comment, marked as answer]
> [cmake output](https://www.toptal.com/developers/hastebin)

I don't see anything there - perhaps you copied a wrong link?

> Which is fed into cbindgen, resulting in the following __internal__imageload.hpp:

Just to be sure: Is it in an `extern "C" block`?

If you run `nm -D yourRustLib.so`, does it show `internal__Load__DynamicImage__FromMemory` as T?
And if you look at the actual failing linker command line, is `yourRustLib.so` 
a) on the line
b) to the right of your cpp image library?

**Comment by IoIxD** [comment]
It was that I had to make an `extern "C"` block, yes.

What's happening now, though, is the linker prioritizes the main function from the Rust code over the one from my C code. Putting a main function within it causes that to be seen as `_start`, and not putting it causes a linker error.

```
/usr/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/14/../../../../lib64/crt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/ligma-browser.dir/build.make:109: ligma-browser] Error 1
make[1]: *** [CMakeFiles/Makefile2:351: CMakeFiles/ligma-browser.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

My CMakeLists.txt is [here](https://raw.githubusercontent.com/IoIxD/ligma-browser/master/CMakeLists.txt). What am I doing wrong?

**Reply by jschwe** [reply]
Why are there two main functions though? If you want to link your Rust code into a C++ executable, then the Rust code should be built as a `staticlib` (or `cdylib` if you need dynamic linking). If you want to link C++ code into a Rust executable, then the C++ code should be built as a static or shared library.

**Reply by IoIxD** [reply]
I am compiling it as a staticlib! The problem is that for some reason corrosion wants a main function anyways.

**Reply by jschwe** [reply]
Corrosion / rust doesn't require a main function for static or dynamic libraries. 
Can you still reproduce the problem if your CMakeLists.txt only contains the the corrosion code for importing the library, and no other code?

**Reply by IoIxD** [reply]
For some reason, making the folder in question it's own `CMakeLists.txt` folder and adding it as a subdirectory worked, something I didn't have to do previously. Weird.
